mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,206 @@
+use crate::{mm_metrics::MmHistogram, MetricType, MetricsJson};
+
+use metrics::{Counter, Gauge, Histogram, Key, KeyName, Label, Recorder, Unit};
+#[cfg(not(target_arch = "wasm32"))]
+use metrics_exporter_prometheus::formatting::{key_to_parts, write_metric_line, write_type_line};
+use metrics_util::registry::{GenerationalAtomicStorage, GenerationalStorage, Registry};
+use std::{collections::HashMap,
+          slice::Iter,
+          sync::{atomic::Ordering, Arc}};
+
+pub struct Snapshot {
+    pub counters: HashMap<String, HashMap<Vec<String>, u64>>,
+    pub gauges: HashMap<String, HashMap<Vec<String>, f64>>,
+    pub histograms: HashMap<String, HashMap<Vec<String>, Vec<f64>>>,
+}
+
+/// `MmRecorder` the core of mm metrics.
+///
+///  Registering, Recording, Updating and Collecting metrics is all done from within MmRecorder.
+pub struct MmRecorder {
+    pub(crate) registry: Registry<Key, GenerationalAtomicStorage>,
+}
+
+impl Default for MmRecorder {
+    fn default() -> Self {
+        Self {
+            registry: Registry::new(GenerationalStorage::atomic()),
+        }
+    }
+}
+
+impl MmRecorder {
+    #[cfg(not(target_arch = "wasm32"))]
+    fn get_metrics(&self) -> Snapshot {
```

Comment:
It turned out to be easier to prepare an example https://github.com/KomodoPlatform/atomicDEX-API/commit/5080da48e81e085ba79f17987a53b8f7039fe2a4 than to explain my idea.
I suggest to avoid `MmRecorder::get_metrics` and `MmRecorder::render` implementations by adding a nested `MmRecorder::prometheus_recorder`.

Pros:
* We avoid complicated implementation (that is actually written by `metrics` team);
* If there is a bug and `metrics` team fix it, we'll just need to update the dependency.

Cons:
* The solution requires 2x memory, because `MmRecorder::registry` and `MmRecorder::prometheus_recorder::Inner::registry` store the same metrics.

Reply:
Not a very good implementation IMO.
Why storing `metrics` into two different registries? there might be thousands of them and this will definitely hurt performance.

from my implementation for rendering `mm_metrics` to `prometheus`, I followed prometheus impl here [https://github.com/metrics-rs/metrics/blob/43287e5855164a0e4931f63ff3efeb533bc74240/metrics-exporter-prometheus/src/recorder.rs#L28-L196](https://github.com/metrics-rs/metrics/blob/43287e5855164a0e4931f63ff3efeb533bc74240/metrics-exporter-prometheus/src/recorder.rs#L28-L196) which only grabs `metrics` from `MmRecorder` Registry and mapped to a `Snapshot` struct which then is use to get all metrics for rendering.

So I think the Cons for your solution is a little bit more tedious than the Pros.

Reply:
I agree with you. Let's check then the implementation :)